Add new infraction filters for the infraction rescheduler

bast0006 opened the pull request:
Add `permanent`, `types`, `expires_before`, and `expires_before` as filter options for the `bot/infractions` endpoint. This let's us filter for only in-flight infractions to expire, and also lets us put limits on how far in the future we reschedule infractions if we desire to.

bast0006 committed:
Add permanent, types, and expires_at filters to bot/infractions endpoint

bast0006 committed:
Add tests for new filters on bot/infractions endpoint

bast0006 committed:
Add a test for ordering infractions returned from the api
As this is used in the new bot infraction rescheduler

bast0006 committed:
Add tests for failure states of `expires_before` and `expires_after` filters

kosayoda reviewed (commented):
From an initial glance, should we have to worry about both the `type` and `types` parameters being used together?

bast0006 committed:
Raise ValidationError for new bot/infractions filter if the types are invalid
If the before time is after the after time, or if both `type` and `types` are specified

bast0006 committed:
Fix bot/infractions after and before filter check being inverted
The check was backwards, enforcing that no results must ever be returned
Relevant tests have been added that actually fetch an infraction with the right times so that I'm sure it's not still broken

bast0006 committed:
Ensure bot/infractions does not accept both expires and permanent filters
Expires and permanent=false are permitted and tested for. Expires_before also filters the database for permanent=false explicitly

kosayoda reviewed (approved):
Looks/tests good to me! Any issues I've missed should surface when we implement the bot side of things.

ChrisLovering reviewed (approved):
Tested and all looks good to me!

Backwards compatible too, so can be merged before the bot PR

ChrisLovering committed:
Merge branch 'main' into bast0006-new-infraction-filters
